### PR TITLE
feat: adds REINFORCE algorithm

### PR DESCRIPTION
**Pull request by abukharin3**
# What does this PR do ?

Adds the REINFORCE algorithm.

# Changelog 
- Please update the [CHANGELOG.md](/CHANGELOG.md) under next version with high level changes in this PR.

# Usage

NAME="2p_reinforce"

# PARAMETERS
RM_NEMO_FILE="/path/to/trained_rm.nemo"

ACTOR_NEMO_FILE="/path/to/sft_model.nemo"

TRAIN_DATA_PATH="/path/to/train_prompts.jsonl"
VALID_DATA_PATH="/path/to/test_prompts.jsonl"

RESULTS_DIR="/path/to/results_dir"
mkdir -p $RESULTS_DIR

GPFS="/path/to/nemo-aligner-repo"
MOUNTS="--container-mounts=MOUNTS" # mounts

CONTAINER=<<<CONTAINER>>> # use the latest NeMo Training container, Aligner will work there

PROJECT=reinforce_run

CRITIC_LOG_DIR="${RESULTS_DIR}/critic_results"
CRITIC_OUTFILE="${CRITIC_LOG_DIR}/critic_output_%j_%t.log"
CRITIC_ERRFILE="${CRITIC_LOG_DIR}/critic_error_%j_%t.err"
REWARD_PORT=5567
CRITIC_CONFIG_PATH="${GPFS}/examples/nlp/gpt/conf"
CRITIC_CONFIG_NAME="inference_rm"

CONF_DIR="${GPFS}/examples/nlp/gpt/conf"
CONFIG_NAME="gpt_reinforce_actor"

mkdir -p $CRITIC_LOG_DIR

CRITIC_NAME="${NAME}_critic"

read -r -d '' cmd_critic_inference <<EOF
cd ${GPFS} \
&& export PYTHONPATH="${GPFS}:${PYTHONPATH}" \
&& export HYDRA_FULL_ERROR=1 \
&& python -u examples/nlp/gpt/serve_reward_model.py \
  --config-path=${CRITIC_CONFIG_PATH} \
  --config-name=${CRITIC_CONFIG_NAME} \
  trainer.num_nodes=1 \
  trainer.devices=8 \
  ++model.tensor_model_parallel_size=4 \
  rm_model_file=${RM_NEMO_FILE} \
  inference.port=${REWARD_PORT}
EOF

srun --het-group=0 -o $CRITIC_OUTFILE -e $CRITIC_ERRFILE --container-image=${CONTAINER} $MOUNTS bash -c "${cmd_critic_inference}" &

sleep 30

ACTOR_LOG_DIR="${RESULTS_DIR}/actor_results"
CHECKPOINT_DIR="${ACTOR_LOG_DIR}/checkpoints"
TENSOBOARD_DIR="${ACTOR_LOG_DIR}/tensorboard"

NUM_ROLLOUTS=16
NORMALIZE="True"
ACTOR_LR="1e-6"
ACTOR_GBS=16
KL=0.01
USE_FLASK=False

mkdir -p $ACTOR_LOG_DIR
mkdir -p $TENSOBOARD_DIR
mkdir -p $CHECKPOINT_DIR

ACTOR_NAME="${NAME}_actor"

host_reward="$(scontrol show hostnames=$SLURM_JOB_NODELIST_HET_GROUP_0 | head -n1)"

read -r -d '' cmd_reinforce <<EOF
cd ${GPFS} \
export PYTHONPATH="${GPFS}:${PYTHONPATH}" \
&& export HYDRA_FULL_ERROR=1 \
&& python -u examples/nlp/gpt/train_gpt_reinforce_actor.py \
  "model.data.data_prefix={train: [${TRAIN_DATA_PATH}], validation: [${VALID_DATA_PATH}], test: [${VALID_DATA_PATH}]}" \
  pretrained_checkpoint.restore_from_path=\"${PRETRAINED_ACTOR_NEMO_FILE}\" \
  exp_manager.checkpoint_callback_params.save_top_k=1 \
  exp_manager.explicit_log_dir=\"${RESULTS_DIR}\" \
  trainer.reinforce.max_epochs=1 \
  trainer.reinforce.max_steps=313 \
  trainer.reinforce.val_check_interval=4 \
  trainer.num_nodes=1 \
  trainer.devices=8 \
  trainer.reinforce.trt_llm.enable=True \
  trainer.reinforce.trt_llm.reshard=True \
  trainer.reinforce.trt_llm.unload_engine_train=False \
  ++model.tensor_model_parallel_size=4 \
  ++model.reinforce.num_rollout_samples=${NUM_ROLLOUTS} \
  model.global_batch_size=${ACTOR_GBS} \
  model.micro_batch_size=1 \
  model.optim.lr=\"\\\$\{multiply:${ACTOR_LR},1.001\}\" \
  model.optim.sched.warmup_steps=0 \
  model.optim.sched.constant_steps=312 \
  model.optim.sched.min_lr=${ACTOR_LR} \
  model.optim.weight_decay=0.01 \
  model.reinforce.rollout_micro_batch_size=16 \
  model.reinforce.forward_micro_batch_size=16 \
  model.reinforce.val_rollout_micro_batch_size=8 \
  model.data.data_impl=jsonl \
  remote_rm.reward_model.ip=${host_reward} \
  remote_rm.reward_model.port=${REWARD_PORT} \
  ++model.reinforce.length_params.max_length=2048 \
  trainer.reinforce.initial_policy_kl_penalty="${KL}" \
  ++model.optim.bucket_cap_mb=200 \
  ++model.dist_ckpt_format=zarr \
  ++model.optim.overlap_grad_sync=False \
  ++model.optim.contiguous_grad_buffer=True \
  ++model.enable_nge=True \
  trainer.reinforce.batch_iterator.use_flask=${USE_FLASK} \
  trainer.reinforce.rollout_batch_seq_length=4096
EOF

srun --het-group=1 -o $PPO_OUTFILE -e $PPO_ERRFILE --container-image=${CONTAINER} $MOUNTS bash -c "${cmd_reinforce}" &

wait

# Before your PR is "Ready for review"
**Pre checks**:
- [Y] Make sure you read and followed [Contributor guidelines](/CONTRIBUTING.md)
- [N] Did you write any new necessary tests?
- [Y] Did you add or update any necessary documentation? Make sure to also update the [NeMo Framework User Guide](https://docs.nvidia.com/nemo-framework/user-guide/latest/index.html) which contains the tutorials

# Checklist when contributing a new algorithm
- [Y] Does the trainer resume and restore model state all states?
- [Y] Does the trainer support all parallelism techniques(PP, TP, DP)?
- [Y] Does the trainer support `max_steps=-1` and `validation`?
- [Y] Does the trainer only call APIs defined in [alignable_interface.py](/nemo_aligner/models/alignable_interface.py)?
- [&] Does the trainer have proper logging?

### CODE REVIEW

File: docs/user-guide/reinforce.rst
Comment (zredeaux07) [inline docs/user-guide/reinforce.rst:245]:
```suggestion
It is important to launch all jobs with ``&`` after the srun command to ensure they do not block each other. 
```

File: docs/user-guide/reinforce.rst
Comment (zredeaux07) [inline docs/user-guide/reinforce.rst:111]:
```suggestion
The above command launches the initial and actor server on one node with eight GPUs.
```

File: docs/user-guide/reinforce.rst
Comment (zredeaux07) [inline docs/user-guide/reinforce.rst:47]:
```suggestion
The above example launches the reward model server on eight GPUs and one node. Make sure to change trainer.devices, trainer.num_nodes depending on your model size and scale. Aligner will work on any scale. Also, make sure to tune the trainer.reinforce.inference_micro_batch_size argument. This argument sets the size of the batch the REINFORCE actor is allowed to send to the reward per DP rank.
```

File: docs/user-guide/reinforce.rst
Comment (zredeaux07) [inline docs/user-guide/reinforce.rst:116]:
```suggestion
You can use slurm to launch the two jobs and get them to coordinate together in a full REINFORCE job through the following:
```

File: docs/user-guide/reinforce.rst
Comment (zredeaux07) [inline docs/user-guide/reinforce.rst:243]:
```suggestion
The above script runs the reward model server on one node and the actor on one node.
```

File: docs/user-guide/reinforce.rst
Comment (zredeaux07) [inline docs/user-guide/reinforce.rst:247]:
```suggestion
.. Note::
```

File: docs/user-guide/reinforce.rst
Comment (zredeaux07) [inline docs/user-guide/reinforce.rst:253]:
```suggestion
After you've completed reinforce training, you can serve your model using the `megatron_gpt_eval.py <https://github.com/NVIDIA/NeMo/blob/8cd5f1c8e7d4fed9f4f946028cd02047c5d2296f/examples/nlp/language_modeling/megatron_gpt_eval.py#L4>`__ script from the NeMo codebase to run more rigorous evaluation of your trained model.
```

File: examples/nlp/gpt/train_gpt_reinforce_actor.py
Comment (zredeaux07) [inline examples/nlp/gpt/train_gpt_reinforce_actor.py:1]: Do we want to update the copyright date?

**Review by zredeaux07 — APPROVED** [review]
LGTM! Just found a few minor nits.

File: examples/nlp/gpt/train_gpt_reinforce_actor.py
Comment (terrykong) [reply examples/nlp/gpt/train_gpt_reinforce_actor.py:1]: @ko3n1g can this be automated? at least the correctness && presence of this header?

File: nemo_aligner/models/nlp/gpt/megatron_gpt_reinforce_actor.py
Comment (terrykong) [reply nemo_aligner/models/nlp/gpt/megatron_gpt_reinforce_actor.py:1]: I don't want to block this PR because of this since it works, but this definitely adds tech debt

File: examples/nlp/gpt/conf/gpt_reinforce_actor.yaml
Comment (terrykong) [inline examples/nlp/gpt/conf/gpt_reinforce_actor.yaml:36]: Should we default to `True`? Fewer models would be supported, but at least things would be very fast OOTB

File: examples/nlp/gpt/conf/gpt_reinforce_actor.yaml
Comment (terrykong) [inline examples/nlp/gpt/conf/gpt_reinforce_actor.yaml:42]: should be okay to drop, but please double check

File: examples/nlp/gpt/conf/gpt_reinforce_actor.yaml
Comment (terrykong) [inline examples/nlp/gpt/conf/gpt_reinforce_actor.yaml:152]: nit: comment not adding much

File: nemo_aligner/algorithms/reinforce.py
Comment (terrykong) [inline nemo_aligner/algorithms/reinforce.py:132]: I think you should statically type at least the model. It helps folks who are using editors to jump to definitions

**Review by terrykong — CHANGES_REQUESTED** [review]
Awesome work!

Please add a functional test for reinforce. You may use the ppo one as reference.

You also need to rebase onto `main` to make sure REINFORCE works with the latest state of TRTLLM (v13). 

The biggest thing about this PR that I have concerns is just the huge overlap between the models since almost all the code is shared. I think we should try to create a common class for all models needing generation.

File: docs/user-guide/reinforce.rst
Comment (abukharin3) [reply docs/user-guide/reinforce.rst]: Fixed.

File: docs/user-guide/reinforce.rst
Comment (abukharin3) [reply docs/user-guide/reinforce.rst]: Changed.

File: docs/user-guide/reinforce.rst
Comment (abukharin3) [reply docs/user-guide/reinforce.rst]: Fixed.

File: docs/user-guide/reinforce.rst
Comment (abukharin3) [reply docs/user-guide/reinforce.rst]: Fixed.

File: docs/user-guide/reinforce.rst
Comment (abukharin3) [reply docs/user-guide/reinforce.rst]: Fixed.

File: docs/user-guide/reinforce.rst
Comment (abukharin3) [reply docs/user-guide/reinforce.rst]: Fixed.

File: docs/user-guide/reinforce.rst
Comment (abukharin3) [reply docs/user-guide/reinforce.rst]: Changed reward_host

File: examples/nlp/gpt/train_gpt_reinforce_actor.py
Comment (abukharin3) [reply examples/nlp/gpt/train_gpt_reinforce_actor.py]: Updated the timer.

File: examples/nlp/gpt/conf/gpt_reinforce_actor.yaml
Comment (abukharin3) [reply examples/nlp/gpt/conf/gpt_reinforce_actor.yaml]: Changed to True.

File: examples/nlp/gpt/conf/gpt_reinforce_actor.yaml
Comment (abukharin3) [reply examples/nlp/gpt/conf/gpt_reinforce_actor.yaml]: Dropped!

File: examples/nlp/gpt/conf/gpt_reinforce_actor.yaml
Comment (abukharin3) [reply examples/nlp/gpt/conf/gpt_reinforce_actor.yaml]: Removed the comment.

File: nemo_aligner/algorithms/reinforce.py
Comment (abukharin3) [reply nemo_aligner/algorithms/reinforce.py]: Fixed this.

File: docs/user-guide/reinforce.rst
Comment (abukharin3) [reply docs/user-guide/reinforce.rst]: Removed it from the name.

File: examples/nlp/gpt/train_gpt_reinforce_actor.py
Comment (terrykong) [reply examples/nlp/gpt/train_gpt_reinforce_actor.py:1]: I'll resolve this since @ko3n1g is aware and working on it in parallel

File: nemo_aligner/models/nlp/gpt/megatron_gpt_reinforce_actor.py
Comment (terrykong) [reply nemo_aligner/models/nlp/gpt/megatron_gpt_reinforce_actor.py:1]: Synced offline, will defer to followup PR

**Review by terrykong — CHANGES_REQUESTED** [review]
Thanks for addressing, just leaving a list of the outstanding TODOs:

- [ ] Add functional test (see [PPO as example](https://github.com/NVIDIA/NeMo-Aligner/blob/main/tests/functional/test_cases/ppo-llama3-pp2-reshard))
- [ ] Address DCO
- [ ] Add unit test https://github.com/NVIDIA/NeMo-Aligner/pull/357#discussion_r1829783436

File: nemo_aligner/utils/ppo_utils.py
Comment (abukharin3) [reply nemo_aligner/utils/ppo_utils.py:117]: Yes, added.

File: docs/user-guide/reinforce.rst
Comment (phtran8) [inline docs/user-guide/reinforce.rst:1]: For line 33, CHECKPOINT_NEMO_FILE should be changed to RM_NEMO_FILE

Comment (abukharin3) [reply docs/user-guide/reinforce.rst:1]: Thanks, fixed.

**Review by terrykong — APPROVED** [review]

**Pull request merged**
